### ISSUE TITLE
blender-9999: segfault `clear_id_nodes()`

### ISSUE THREAD

**Issue by LinuxUserGD** [brothermechanic/cg#72]
```bash
media-gfx/blender-9999::cg was built with the following:
USE="X abi11-compat alembic avi boost bullet cg collada color-management cycles draco embree ffmpeg fftw flac fluid freestyle gmp icu jpeg2k libmv llvm lzma mold mp3 nls obj oceansim oidn openal openexr openmp openpgl opensubdiv openvdb openxr opus osl pdf ply potrace pugixml pulseaudio quadriflow quicktime remesh sdl simd smoke stl tbb theora vorbis vpx vulkan wayland webp x264 xvid -abi10-compat -abi7-compat -abi8-compat -abi9-compat -aom (-cuda) -cuda_targets_compute_89 -cuda_targets_sm_30 -cuda_targets_sm_35 -cuda_targets_sm_37 -cuda_targets_sm_50 -cuda_targets_sm_52 -cuda_targets_sm_60 -cuda_targets_sm_61 -cuda_targets_sm_70 -cuda_targets_sm_75 -cuda_targets_sm_86 -cuda_targets_sm_89 -cycles-bin-kernels -debug -doc -dpx -gold -gtests -headless -jack -jemalloc -lld -lzo -man -materialx -nanovdb -ndof -oneapi (-optix) -portable -rocm -sndfile -test -usd -valgrind" AMDGPU_TARGETS="-gfx1010 -gfx1011 -gfx1012 -gfx1030 -gfx1031 -gfx1032 -gfx1034 -gfx1035 -gfx1100 -gfx1101 -gfx1102 -gfx900 -gfx902 -gfx90c" L10N="de -ab -ar -be -bg -ca -cs -da -el -en -eo -es -es_ES -eu -fa -fi -fr -ha -he -hi -hr -hu -id -it -ja -ka -km -ko -ky -ne -nl -pl -pt -pt_BR -ru -sk -sr -sr@latin -sv -sw -ta -th -tr -uk -vi -zh_CN -zh_HANS -zh_HANT -zh_TW" PYTHON_SINGLE_TARGET="python3_12 -python3_10 -python3_11"
FEATURES="ccache sfperms news usersandbox unknown-features-warn usersync unmerge-logs binpkg-dostrip unmerge-orphans network-sandbox config-protect-if-modified distlocks buildpkg-live qa-unresolved-soname-deps preserve-libs binpkg-docompress assume-digests sandbox pid-sandbox merge-sync protect-owned userfetch xattr fixlafiles parallel-fetch ebuild-locks binpkg-multi-instance binpkg-logs ipc-sandbox userpriv pkgdir-index-trusted strict buildpkg"
```

I get the following segfault when running blender compiled with llvm slot 17 and musl libc:

```bash
user@gentoo ~ $ sudo gdb --args blender
GNU gdb (Gentoo 14.1 vanilla) 14.1
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-gentoo-linux-musl".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from blender...
(No debugging symbols found in blender)
(gdb) run
Starting program: /usr/bin/blender 
[New LWP 15372]
[New LWP 15373]
[New LWP 15374]
[New LWP 15375]
GHOST/Wayland: error: XDG_RUNTIME_DIR is invalid or not set in the environment.
0x5555581f1d72 <BLI_system_backtrace+0x32> at /usr/bin/blender
0x7fffe3a568f2 <???> at /lib/libwayland-client.so.0
0x7fffe3a52c87 <wl_display_connect+0x2b7> at /lib/libwayland-client.so.0
0x55555a460030 <_ZN19GHOST_SystemWaylandC1Eb+0x150> at /usr/bin/blender
0x55555a449895 <_ZN13GHOST_ISystem12createSystemEbb+0xa5> at /usr/bin/blender
0x5555582e2320 <_Z13wm_ghost_initP8bContext+0x60> at /usr/bin/blender
0x55555829f5f2 <_Z7WM_initP8bContextiPPKc+0x452> at /usr/bin/blender
[Detaching after vfork from child process 15376]
[New LWP 15379]
[LWP 15379 exited]
[New LWP 15380]
[New LWP 15381]
[New LWP 15382]
[New LWP 15383]
[New LWP 15387]
[New LWP 15388]
[New LWP 15389]
[New LWP 15390]
[New LWP 15391]

Thread 1 "blender" received signal SIGSEGV, Segmentation fault.
0x00005555581faa7c in blender::deg::Depsgraph::clear_id_nodes() ()
(gdb) 
```
https://github.com/blender/blender/blob/main/source/blender/depsgraph/intern/depsgraph.cc#L156-L174

Because blender doesn't officially support musl libc, I've applied the following patch to compile blender:
```bash
diff --git a/source/blender/freestyle/intern/geometry/VecMat.h b/source/blender/freestyle/intern/geometry/VecMat.h
index ba5ad84a156..3b0c2ec5a81 100644
--- a/source/blender/freestyle/intern/geometry/VecMat.h
+++ b/source/blender/freestyle/intern/geometry/VecMat.h
@@ -36,6 +36,9 @@ template<> struct is_false<false> {
 //
 /////////////////////////////////////////////////////////////////////////////
 
+#define _GNU_SOURCE 1
+typedef unsigned u_int, uint;
+
 template<class T, uint N> class Vec {
  public:
   typedef T value_type;
diff --git a/source/blender/freestyle/intern/system/Id.h b/source/blender/freestyle/intern/system/Id.h
index 6f0d2b3c4f4..f266f2feaad 100644
--- a/source/blender/freestyle/intern/system/Id.h
+++ b/source/blender/freestyle/intern/system/Id.h
@@ -20,6 +20,8 @@ namespace Freestyle {
  */
 class Id {
  public:
+  #define _GNU_SOURCE 1
+  typedef unsigned u_int, uint;
   typedef uint id_type;
 
   /** Default constructor */
diff --git a/source/blender/freestyle/intern/geometry/Geom.h b/source/blender/freestyle/intern/geometry/Geom.h
index 53104aa792a..5dd0911fd55 100644
--- a/source/blender/freestyle/intern/geometry/Geom.h
+++ b/source/blender/freestyle/intern/geometry/Geom.h
@@ -17,6 +17,9 @@ namespace Freestyle {
 
 namespace Geometry {
 
+#define _GNU_SOURCE 1
+typedef unsigned u_int, uint;
+
 typedef VecMat::Vec2<uint> Vec2u;
 typedef VecMat::Vec2<int> Vec2i;
 typedef VecMat::Vec2<float> Vec2f;
```
make.conf
```bash
COMMON_FLAGS="-O3 -pipe -march=native -mtune=native -D_FORTIFY_SOURCE=3 -flto=thin -g0"
CC="clang"
CXX="clang++"
HOSTCC="clang"
HOSTCXX="clang++"
CFLAGS="${COMMON_FLAGS}"
CXXFLAGS="${COMMON_FLAGS} -stdlib=libc++"
FCFLAGS="${COMMON_FLAGS}"
FFLAGS="${COMMON_FLAGS}"

LDLIBS="-L/use/lib -lexecinfo"
LDFLAGS="${COMMON_FLAGS} ${LDLIBS} -Wl,-O3 -Wl,--as-needed -Wl,--strip-debug -Wl,--undefined-version -Wl,--icf=safe -Wl,--threads=4 -Wl,--compress-debug-sections=zstd -fuse-ld=lld -rtlib=compiler-rt -unwindlib=libunwind"
CHOST="x86_64-gentoo-linux-musl"
```

**Comment by anex5** [comment]
First, I can see following error in the log file you provide: GHOST/Wayland: error: XDG_RUNTIME_DIR  is invalid or not set in the environment. Maybe this error is not related to the segfault, but it is definitely an error. You should start sway (or other WM you use) with script, which provides XDG_RUNTIME_DIR. For example:
```
if test -z "${XDG_RUNTIME_DIR}"; then
    export XDG_RUNTIME_DIR=/tmp/${UID}-runtime-dir
    if ! test -d "${XDG_RUNTIME_DIR}"; then
        mkdir -p "${XDG_RUNTIME_DIR}"
        chmod 0700 "${XDG_RUNTIME_DIR}"
    fi
fi
```
Also I would recommend you to build blender without LTO i.e. put `-fno_lto` and  `-O2`  in `COMMON_FLAGS` variable in your make.conf file.

**Comment by LinuxUserGD** [comment]
> Also I would recommend you to build blender without LTO i.e. put -fno_lto and -O2 in COMMON_FLAGS variable in your make.conf file.

Thanks, `blender-4.1.0` and `blender-9999` segfaults but `blender-4.0.2` compiled with the following user patches after disabling `osl`:
- [blender.patch](https://gist.github.com/LinuxUserGD/32d9bdad8378fe3f6a6f45aee2753d9e#file-blender-patch)
- blender-3.3.0-fix-build-with-boost-1.81.patch
- blender-3.3.6-cycles-gcc13.patch
- blender-4.0.1-fix-cflags-cleaner.patch
